### PR TITLE
feat(ddm): Add support for blocking metrics

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR adds the ability to store blocked metrics in sentry options and allows to return of those blocked metrics in the `metrics/meta` endpoint.

A follow-up PR will be done in which the metrics configuration will be emitted for Relay. It was not done in this PR to not add additional complexity and surface for errors.

### CODE REVIEW

File: src/sentry/api/endpoints/organization_metrics.py [hunk 1]
```diff
@@ -50,6 +55,45 @@ def get_use_case_id(request: Request) -> UseCaseID:
 
 @region_silo_endpoint
 class OrganizationMetricsEndpoint(OrganizationEndpoint):
+    publish_status = {"GET": ApiPublishStatus.EXPERIMENTAL, "PATCH": ApiPublishStatus.EXPERIMENTAL}
+    owner = ApiOwner.TELEMETRY_EXPERIENCE
+
+    def get(self, request: Request, organization) -> Response:
```
Comment (iambriccardo) [inline src/sentry/api/endpoints/organization_metrics.py:61]: Moved the `metrics/meta` endpoint also here, to prepare for a future refactor.

File: src/sentry/api/urls.py [hunk 2]
```diff
@@ -1952,10 +1953,15 @@ def create_group_urls(name_prefix: str) -> list[URLPattern | URLResolver]:
         name="sentry-api-0-organization-ddm-meta",
     ),
     re_path(
-        r"^(?P<organization_slug>[^/]+)/metrics/meta/$",
+        r"^(?P<organization_slug>[^/]+)/metrics/$",
         OrganizationMetricsEndpoint.as_view(),
         name="sentry-api-0-organization-metrics-index",
     ),
+    re_path(
+        r"^(?P<organization_slug>[^/]+)/metrics/meta/$",
```
Comment (iambriccardo) [inline src/sentry/api/urls.py:1961]: This is the previous endpoint, which is kept for backwards compatibility.

File: src/sentry/sentry_metrics/visibility/metrics_blocking.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    @classmethod
+    def load_from_project(cls, project: Project, repair: bool = False) -> "BlockedMetrics":
+        json_payload = project.get_option(BLOCKED_METRICS_PROJECT_OPTION_KEY)
+        if not json_payload:
+            return BlockedMetrics(metrics=[])
+
+        try:
+            blocked_metrics_payload = json.loads(json_payload)
+        except ValueError:
+            if repair:
```
Comment (iambriccardo) [inline src/sentry/sentry_metrics/visibility/metrics_blocking.py:51]: Repairing on write allows us to make sure we don't end up in a broken state of the system which requires a separate PR to unbreak.

File: src/sentry/snuba/metrics/datasource.py [hunk 4]
```diff
@@ -188,32 +192,84 @@ def get_available_derived_metrics(
...
-                mri=metric_mri,
-                name=parsed_mri.name,
-                operations=cast(
-                    Sequence[MetricOperationType], get_available_operations(parsed_mri)
-                ),
-                unit=cast(MetricUnit, parsed_mri.unit),
-                type=parsed_mri.entity,
-                project_ids=project_ids,
+        blocked_for_projects = []
+        if (blocked := blocked_metrics.get(metric_mri)) is not None:
```
Comment (iambriccardo) [inline src/sentry/snuba/metrics/datasource.py:242]: Too much manipulation going on here but if we have to merge blocked and non-blocked metrics we have no other ways.

File: src/sentry/snuba/metrics/datasource.py [hunk 5]
```diff
@@ -188,32 +192,84 @@ def get_available_derived_metrics(
...
+
+
 def get_metrics_meta(projects: Sequence[Project], use_case_id: UseCaseID) -> Sequence[MetricMeta]:
-    metas = []
-    stored_mris = get_stored_mris(projects, use_case_id) if projects else {}
+    if not projects:
+        return []
 
-    for metric_mri, project_ids in stored_mris.items():
+    stored_metrics = get_stored_metrics_of_projects(projects, use_case_id)
```
Comment (iambriccardo) [inline src/sentry/snuba/metrics/datasource.py:232]: driveby: Removed for consistency.

File: src/sentry/sentry_metrics/visibility/metrics_blocking.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+BLOCKED_METRICS_PROJECT_OPTION_KEY = "sentry:blocked_metrics"
+
+
+class BlockedMetricsRelayConfig(TypedDict):
+    deniedNames: Sequence[str]
+
+
+@dataclass(frozen=True)
+class BlockedMetric:
+    metric_mri: str
```
Comment (obostjancic) [inline src/sentry/sentry_metrics/visibility/metrics_blocking.py:19]: nit: 
```suggestion
    mri: str
```

File: src/sentry/sentry_metrics/visibility/metrics_blocking.py (hunk 3 above)
Comment (obostjancic) [reply src/sentry/sentry_metrics/visibility/metrics_blocking.py:51]: I do understand that this is a real possibility since anyone else could theoretically write to this field, but I do not see a value in trying to "repair" it by deleting the key completely. Raising an exception and opening an issue on Sentry would be enough IMO.

File: src/sentry/sentry_metrics/visibility/metrics_blocking.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        for blocked_metric in blocked_metrics_payload:
+            blocked_metric_obj = BlockedMetric.from_dict(blocked_metric)
+            # We want to implement a best effort mechanism in which we will notify Sentry in case of invalid blocked
+            # metrics, but this won't have effect on the other metrics.
+            if blocked_metric_obj is not None:
+                blocked_metrics.append(blocked_metric_obj)
+
+        return BlockedMetrics(metrics=blocked_metrics)._merge_blocked_metrics()
+
+    def _merge_blocked_metrics(self) -> "BlockedMetrics":
```
[the remark this answers is not in the document]
Comment (obostjancic) [reply src/sentry/sentry_metrics/visibility/metrics_blocking.py:76]: We need merging because of the tags?

File: src/sentry/snuba/metrics/datasource.py [hunk 8]
```diff
@@ -1,5 +1,7 @@
 from __future__ import annotations
 
+from sentry.sentry_metrics.visibility import get_blocked_metrics
```
Comment (obostjancic) [inline src/sentry/snuba/metrics/datasource.py:3]: nit: do we plan on putting anything else in this module? If not, then i would call it `blocking` or `metrics_blocking` or would flatten it out by butting exports and errors in `metrics_blocking.py`

**Review by obostjancic — APPROVED** [review]
Looks fine overall, left a couple of comments. Also, as mentioned before, I think we should try to keep refactors to a minimum during feature impl.

File: src/sentry/sentry_metrics/visibility/metrics_blocking.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (iambriccardo) [reply src/sentry/sentry_metrics/visibility/metrics_blocking.py:76]: Nope, we need merging in general, since we don't want to return to the frontend duplicated data.

File: src/sentry/sentry_metrics/visibility/metrics_blocking.py (hunk 3 above)
Comment (iambriccardo) [reply src/sentry/sentry_metrics/visibility/metrics_blocking.py:51]: I can remove this behavior, but if something is corrupted, we either delete it or figure out its content and make a PR to untangle it programmatically.

My main idea for doing so is that if a config of a user gets corrupted, they will now ingest all their blocked metrics and they don't have a way to block them again (they have to wait for our PR). With my system, they can immediately fix the problem with the downside being losing their previous settings. Still, it's a negligible problem since if the JSON is corrupted there likely won't be valuable data in the stored settings.

File: src/sentry/snuba/metrics/datasource.py (hunk 8 above)
Comment (iambriccardo) [reply src/sentry/snuba/metrics/datasource.py:3]: Yes we are, all the deletion logic will go there. It's called visibility exactly because it's a generic term that both have in common.
